### DISCUSSION TITLE
How would i add more options to the options menu

### DISCUSSION THREAD

**Discussion by Raptorq1** [Techokami/SonicWorldsNext#142] [Q&A]
How would i add more options to the options menu

**Comment by r6915ee** [comment]
What kind of options? If it's a controls menu, then that is present already; you'll just need to press F3 to open and close it.

**Reply by Raptorq1** [reply]
Sound and video options

**Reply by r6915ee** [reply]
So something like an added voice volume option as an example?

**Reply by Raptorq1** [reply]
Yes

**Reply by r6915ee** [reply]
If it's just a voice volume option or another volume option, you'll just most likely have to go through the regular process of doing it in Godot and adding it to Global.gd.

For making the audio controllable on its own, you'll need to go over to the bottom toolbar, and press on Audio. There, to have your voice volume controllable, you will need to click on "Add Bus." This is ESSENTIAL for having the voice volume being able to be controlled separately from every other audio bus. Rename it to whatever your specific audio bus is supposed to be, for example, "Voice."

Then, you should go on over to the save_settings() function definition in Global.gd, easily accessible through the Autoload tab in project settings. There, copy and paste either one of the top two functions used (the SFX and Music ones) and replace the name of those (EXCEPT Volume) with your audio bus name.

Afterwards, head on over to load_settings(); it's just right next to save_settings(). It's just if statements, but you should be able to do the same thing before here but with the if statements.

From the looks of it, Sonic Worlds Next does NOT have an array or something to hold all of the options data entries. It seems to rather get them all from Settings.cfg, so you don't need to worry about adding a new entry manually. I'm not sure if this is the case, though, as I cannot find any mentions of said array within the source code.

Afterwards, it's a bit on your own. You may add it as a new node through the use of a duplicate of the soundChannel variable in Global.gd in case you want to have voices during gameplay, or have it just be used as its own variable in each of your scripts and AudioStreamPlayers using it. This one's up to you.

You CAN use the first step in other projects, because it is a Godot feature on its own. So if you want to have volume options all across multiple projects that use different audio systems, you may use it and do the specific steps for the specified project afterwards using your new audio bus.

**Reply by Raptorq1** [reply]
I didn't mean  voice menu  options

**Reply by r6915ee** [reply]
Then what do you specifically mean then in terms of options?

**Reply by Raptorq1** [reply]
I will give you the code instead
[SoundManager.zip](https://github.com/Techokami/SonicWorldsNext/files/14381284/SoundManager.zip)

**Reply by r6915ee** [reply]
From what I'm understanding what you want to do, if you were to want to add new options for the script you gave, you would have to add a new audio bus for each sound option you want to be changeable (in case it's controllable like SFX) and add them to your menusText array and modify lines 54-55 to also include the new changes.

If you want to directly implement certain things like the remastered music or the music styles, that would certainly be a tad bit tricky if not done right. It might be possible to check in the settings what the settings are currently set to inside of Global.gd, having the settings' current values tell it in which directory or filename system or something you have to identify what music file to look for. This would also conclude in rewriting a bit of the themes system in Global.gd, however.

**Reply by Raptorq1** [reply]
I all ready have a music style options setter

**Reply by r6915ee** [reply]
Then what do you exactly need?

**Reply by Raptorq1** [reply]
Implementing remastered music option

**Comment by r6915ee** [comment]
> If you want to directly implement certain things like the remastered music or the music styles, that would certainly be a tad bit tricky if not done right. It might be possible to check in the settings what the settings are currently set to inside of Global.gd, having the settings' current values tell it in which directory or filename system or something you have to identify what music file to look for. This would also conclude in rewriting a bit of the themes system in Global.gd, however.

Then you should probably do what I've put above. In case you have multiple directories for the regular music and the remastered music separate, you could probably replace the music option in a level with a corresponding ID to the song file (such as the numbers listed at the beginning of the file name). It would require a tad bit of rewriting for the level's music system and also Global.gd's themes system, of course, but it should work.
